For N=11
Numerical Integral I  = 1.8932978307957875
Analytic Solution     = 1.9052386904826757
Relative Error        = -6.2673825314050191e-03






For N=101 Numerical Integral I  = 1.9051192179747314
Analytic Solution     = 1.9052386904826757
Relative Error        = -6.2707370231951387e-05 





For N=1001
Numerical Integral I  = 1.9052374957512166
Analytic Solution     = 1.9052386904826757
Relative Error        = -6.2707705084097398e-07





For N=10001
Numerical Integral I  = 1.9052386785353552
Analytic Solution     = 1.9052386904826757
Relative Error        = -6.2707735670741727e-09





For N=10000000000000000000000000000000000001
Numerical Integral I  = -0.3926990816987243
Analytic Solution     = 1.9052386904826757
Relative Error        = -1.2061154246239023e+00

Q1: The integral is analytic and its solution is Itrue, how far are I from real solution?




Ans: I tried increasing value of N by 10 times and every time you increase value of N by 10 times, the relative error decrease by 100 times. for very high value of N, this error will become negligible or zero. 

Q2: How can you reduce the relative error?

Ans: By refining the grid or increasing number of N.


Q3: Minimum relative error you can find using those technique.


Ans: Using the composite trapezoid rule, I observed the clear convergence. While increasing the value of N, relative error was also reduced. the minimum relative error was measured from trapezoid rule was 6.27E10^-9 at N=10001. 

Q4:   Using the output file you produce, use an interpreted language to calculate the same integral (call it I4).

In [1]:
import numpy as np
import math

# 1) Load the sampled data the C code wrote: columns = x, f(x)
x, fx = np.loadtxt("task4_output.txt", unpack=True)

# 2) Trapezoidal rule on the same grid
#    (matches exactly what your C code did)
h = x[1] - x[0]
I4 = h*(0.5*fx[0] + fx[1:-1].sum() + 0.5*fx[-1])

# 3) If you want, also check vs the analytic value:
I_true = (math.exp(math.pi/2) - 1.0)/2.0
abs_err_vs_true = abs(I4 - I_true)

# 4) Put here the I printed by your C program for the SAME N
I_from_C = 1.9052374957512166   # <-- e.g., your N=1001 result
absrel = abs(I_from_C - I4)

print(f"I4 from file           = {I4:.16f}")
print(f"abs error |I - I4|     = {absrel:.16e}")
print(f"(Optional) |I4 - Itrue|= {abs_err_vs_true:.16e}")


I4 from file           = 1.8932977983971142
abs error |I - I4|     = 1.1939697354102385e-02
(Optional) |I4 - Itrue|= 1.1940892085561483e-02


Using the output text file and python code, the obtained I4 is 1.8932977983971142. The absolute error compared with the C results is I-I4=1.19E10^-2. The two output values are consistent but not very close to the analytic solution. 